server/src/main/java/io/crate/types/IntEqQuery.java

```diff
@@ -48,11 +59,14 @@ public Query rangeQuery(String field,
         if (upperTerm != null) {
             upper = includeUpper ? upperTerm.intValue() : upperTerm.intValue() - 1;
         }
-        Query indexquery = IntPoint.newRangeQuery(field, lower, upper);
-        if (hasDocValues) {
-            return new IndexOrDocValuesQuery(
-                    indexquery, SortedNumericDocValuesField.newSlowRangeQuery(field, lower, upper));
+        boolean isIndexed = indexType != IndexType.NONE;
+        if (hasDocValues && isIndexed) {
+            return IntField.newRangeQuery(field, lower, upper);
+        } else if (hasDocValues) {
+            return SortedNumericDocValuesField.newSlowRangeQuery(field, lower, upper);
+        } else if (isIndexed) {
```

Comment:
Please keep using `IndexOrDocValuesQuery` if both doc-values and inverted index is available

Reply:
Thank you changed it, but `IndexOrDocValuesQuery` still causes `TooManyClauses` exception from `testWhereNotEqualAnyWithLargeArray` https://github.com/crate/crate/pull/14527#discussion_r1285121912. It could be because `IndexOrDocValuesQuery` is experimental, do you have any suggestion for this?

Btw, if we update `testWhereNotEqualAnyWithLargeArray` from `id != any(` to `id < any(` the same exception will happen on master too.

Reply:
Since we state 

https://crate.io/docs/crate/reference/en/5.4/config/node.html#queries
> This setting defines the maximum number of elements an array can have ...

can we validate the number of elements in `refMatchesAnyArrayLiteral` and increase `indices.query.bool.max_clause_count` say 3 times larger to workaround `TooManyClauses` exception?

Another option could be to use either one of index or doc-value query. cc @matriv

Reply:
Looks like I missed in the diff that we didn't use `IndexOrDocValuesQuery` before in this case.

Fine to leave it out for now. 

---

For future reader: The problem is that in `ANY !=` we create a BooleanQuery with a MUST clause per value.
The must clause uses a `termQuery` for the inner value and if we use `IndexOrDocValuesQuery` it counts as two instead of one - so we end up multiplying the number of clausess.

Reply:
Thx @mfussenegger, we clarified this offline @jeeminso, no? On top https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/IndexOrDocValuesQuery.java#L54 it's experimental, so we shouldn't use it, for now at least.

Reply:
It just means the API might change in incompatible ways in the future. No reason not to use it if it works and brings performance improvements.

See also https://github.com/apache/lucene/issues/10929

Reply:
Didn't know, thx!

Reply:
Yes we did @matriv, thank you, just wanted to make sure everyone involved are on the same page.

Reply:
My understanding is that if we update `testWhereNotEqualAnyWithLargeArray` from termQuery to rangeQuery (from `id != any(` to `id < any(`) the same exception will be thrown, so `IndexOrDocValuesQuery` also cannot be used for rangeQueries.

Btw, adding more variations of that test will be handled by https://github.com/crate/crate/pull/14604.

Reply:
Ah you're right, thanks for the clarification.